### ISSUE TITLE
tracking rc3 compliance

### ISSUE THREAD

**Issue by tkellen** [endpoints/endpoints#94]
/ref json-api/json-api#484

**Comment by bobholt** [comment]
I made a gist of the diff between RC2 and RC3: https://gist.github.com/bobholt/1a5e9103be5fa85a53da

**Comment by bobholt** [comment]
This implementation for this section will need to change rather drastically: http://jsonapi.org/format/#document-structure-resource-relationships

**Comment by tkellen** [comment]
Once that #96 lands, are we in a position to start actually using this?

@bmac, what's the scoop on our ember-data adapter?

@bobholt, can you let me know when we're ready so I can take a minute to figure out what's needed to confirm v2 of our api running off of a specific sha of endpoints?

**Comment by bobholt** [comment]
@tkellen. Maybe? I want to take another pass through the spec to make sure we got all of the required bits. We'll still be short pagination and extensions, I think.

Now go away until after your vacation!

![Harry](http://s.mcstatic.com/thumb/7626259/20786404/4/flash_player/0/1/harry_and_the_hendersons_1987_saying_goodbye_part_2.jpg)

**Comment by tkellen** [comment]
Hahahahaha, I'm still doing minor email triage in the morning and before bed. This vacation won't be worth it if I come home to 40,000 emails. We spent all day yesterday doing scuba training, and then dove to 40 feet for about an hour--it was AMAZING!

Off to do it again--happy hacking :)

**Comment by bobholt** [comment]
On hold pending resolution of https://github.com/json-api/json-api/issues/482

**Comment by bobholt** [comment]
Remaining conformance tasks after #96. Note these are all **MUST**s in the spec; optional features haven't been catalogued here:

### General
- [ ] `it must ignore request object members not recognized by the specification`
- [ ] `it must represent a logical collection of resources as an array, even if it only contains one item`
- [ ] `it must represent a logically singular resource as a single resource object`

### Create
- [ ] `it must return 403 Forbidden in response to an unsupported request using a client-generated ID`
  - endpoints currently allows all requests using client-generated IDs. We would need global configuration options as described in #51 to allow endpoints to optioanlly not support client-generated IDs
  - test not implemented
- [ ] `it must return 409 Conflict when processing a request to create a resource with an existing client-generated ID`
  - this requires fixing the TODOs in `/modules/request-handler/lib/verify_data_object.js`
  - tests skipped

### Delete
- [ ] `it must not allow partial updates`
  - this requires bookshelf updates
  - test not implemented

### Read
- [ ] `it must support fetching resource for URLs provided as a `self` link in a links object`
  - test not implemented
- [ ] `it('must support fetching resource for URLs provided as a `related` link as part of a link object`
  - test not implemented
- [ ] `it must respond to a successful request to fetch an individual resource or collection with a 200 OK response`
  - test not implemented
- [ ] `it must respond to a successful request to fetch a resource collection with an array as the document's primary data`
  - test not implemented
- [ ] `it('must respond to a successful request to fetch an individual resource with a resource object or null as the document\'s primary data`
  - test not implemented
- [ ] `it('must return 404 Not Found when processing a request to fetch a resource that does not exist`
  - test not implemented
- [ ] `it must respond to a successful request to fetch a relationship with a 200OK response`
  - test not implemented
- [ ] `it('must have a primary data consisting of null, an object of type and id members, an array`
  - test not implemented
- [ ] `it('must return 404 Not Found when processing a request to fetch a relationship URL that does not exist`
  - test not implemented

### Update
- [ ] `it must not allow partial updates`
  - this requires bookshelf updates
  - test not implemented
- Updating to-One Relationships
  - [ ] `it must require a top-level data member containing either an object with type and id members or null`
    - should return an HTTP error if this check fails
    - not implemented
- Updating to-Many Relationships
  - [ ] `it must respond to POST requests to a to-many relationship URL
    - test not implemented
  - [ ] `it must respond to DELETE requests to a to-many relationship URL`
    - test not implemented
  - [ ] `it must require a top-level data member containing either an object with type and id members`
    - test not implemented
  - [ ] `it must completely replace every member of the relationship on a PATCH request if allowed`
    - test not implemented
  - [ ] `it must return an appropriate error if some resources cannot be found or accessed`
    - test not implemented
  - [ ] `it must return a 403 Forbidden if complete replacement is not allowed by the server`
    - test not implemented
  - [ ] `it must append specified members of a POST request`
    - test not implemented
  - [ ] `it must not add existing type and id combinations again`
    - test not implemented
  - [ ] `it must return a 204 No Content if the resource is successfully added or already present`
    - test not implemented
  - [ ] `it must either DELETE members of the relationship or return 403 Forbidden on a DELETE request`
    - test not implemented
  - [ ] `it must return a 204 No Content if the resource is successfully deleted or is already missing`
    - test not implemented
  - [ ] `it must return 204 No Content if the update is successful and the attributes remain up to date`
    - test not implemented

**Comment by bobholt** [comment]
Alright, @tkellen, I'm tagging out. Leaving 2 things to you that I think are related to Bookshelf limitations:

- Make sure all PATCH/DELETE requests completely process or are rolled back.
- Bring back the validation commented out in the refactor [here](https://github.com/endpoints/endpoints/blob/master/src/request-handler/lib/verify_data_object.js#L31-L49)

**Comment by tkellen** [comment]
Sounds good @bobholt! Thank you **so much** for your help! I hope you have some time to help here and there in the future :)

I'll take a pass at the API today to get v2 running off of master.
